rfc/system/3346-array-node.md

```diff
@@ -0,0 +1,105 @@
+2. Override `NodeExecutionContext` components in the case of a `TaskNode` subtask: A high priority item is ensuring the same level of performance, and perhaps more importantly compatibility, as the existing `map_task` implementation. To these ends, we need to override the `eventRecorder` so that each subtask execution does not send individual events but rather appends an `ExternalResource` to a single `TaskExecutionEvent`, we need to override the `TaskStateWriter` and `TaskStateReader` so that subtask state persistence is minimal, among others. This will be difficult, but will ensure a backwards compatible and performant evaluation. In the case where this is not possible, we can choose to proceed without this level of emulation, resulting in a less performant and backwards incompatible solution.
+3. Execute subtasks using the `RecursiveNodeHandler` call using our mocked and overridden entities. This is a common entry point for all Flyte Node evaluation and ensures good software engineering practices and identical support for Node evaluation in subtasks as all other Flyte Nodes.
+ 
+There are (at least) two very important things to note with this implementation:
+
+**Cache lookups will happen at the TaskHandler level:** This is in parity with the existing `map_task` approach, where cache lookups are batched and performed asynchronously. The purpose for this change is two-fold (1) it ensures no code duplication is necessary to support all caching functionality (ex. cache serialize and cache overwrite) and (2) it ensure compatibility with the unreleased "fast-cache" work, which should improve evaluation of cached Nodes significantly.
+
+**Resolving subtask Node input values is not clearly defined:** Currently, FlytePropeller sets environment variables to inform flytekit which input index the subtask should read for a collection of data values. This ensure performance in that each subtask does not require separate blobstore storage. The main problem is that FlytePropeller needs a single input value to compute the cache key. There are a few potential solutions that we will have to evaluate for viability:
+- Creating a flyteidl `Literal` which supports indexing in to a List, so a subtask Node input will be index N of the Literal in the blobstore.
+- Storing separate inputs in the blobstore. This is very inefficient and should be used as a last resort.
+- Some other fancy solution we hack together in the heat of the moment.
+
+#### flytekit
+
+flytekit will require 3 known updates, but further analysis here will defer to those with more complete context:
+1. Compiling `map_task` definitions into new ArrayNode rather than a TaskNode using the `k8s_array` backend plugin.
+2. Adding support for multiple input values.
```

Comment:
Does this refer to a `partial` or a `zip` approach? I've seen both being implemented in `flytekit` in https://github.com/flyteorg/flytekit/pull/1556 - would be nice to support both for the new `ArrayNode`.

Reply:
Exactly, we are sorting out a few issues with the approach mentioned above (ex. needs `flyteidl` updates to support passing lists as partials). All of this support will be included in the proposed `ArrayNode` implementation.